ecc/bw6-761/pairing.go

```diff
@@ -640,3 +640,97 @@ func (p *G2Affine) tangentCompute(evaluations *LineEvaluationAff) {
 	evaluations.R1.Mul(&λ, &p.X).
 		Sub(&evaluations.R1, &p.Y)
 }
+
+// ------------------------
+// direct-extension pairing
+// ------------------------
+
+// MillerLoopDirect computes the multi-Miller loop using the direct E6 extension
+// and returns a towered E6 element. This version corresponds to gnark circuit.
+// ∏ᵢ MillerLoop(Pᵢ, Qᵢ) =
+// ∏ᵢ { fᵢ_{x₀+1+λ(x₀³-x₀²-x₀),Qᵢ}(Pᵢ) }
+//
+// Alg.2 in https://eprint.iacr.org/2021/1359.pdf
+// Eq. (6') in https://hackmd.io/@gnark/BW6-761-changes
+func MillerLoopDirect(P []G1Affine, Q []G2Affine) (GT, error) {
+	lines := make([][2][len(LoopCounter) - 1]LineEvaluationAff, 0, len(Q))
+	for _, qi := range Q {
+		lines = append(lines, PrecomputeLines(qi))
+	}
+
+	// check input size match
+	n := len(P)
```

Comment:
Would move this sanity check part before the line precomputation. Then you can just measure `n != len(Q)`. Then we fail early before computing the lines.